cheatsheets/Clickjacking_Defense_Cheat_Sheet.md

```diff
@@ -311,3 +313,166 @@ Activate [designMode](https://developer.mozilla.org/en-US/docs/Web/API/Document/
+        }
+    });
+}, { threshold: 0.5 });
+
+document.querySelectorAll("iframe").forEach(iframe => observer.observe(iframe));
+```
+
+✅ Prevents hidden iframe overlays from capturing clicks.
+
+✅ Ensures users interact only with visible elements.
+
+#### **2. Enforce Click-Only on Trusted Domains**
+
+**Concept:** Attackers host iframes on malicious domains. This script ensures actions only occur if the request comes from a trusted domain.
+
+```javascript
+if (document.referrer && !document.referrer.includes("your-website.com")) {
+    alert("Warning: Clicks from an untrusted source detected!");
+}
+```
+
+✅ Blocks interactions initiated from untrusted sites.
+
+✅ Prevents attacks executed via external embedding.
+
+### **3. Delayed Click Action (Click Delay)**
+
+**Concept:** Prevents rapid unintended double-clicks by enforcing a minimum delay between clicks.
+
+```javascript
+let lastClickTime = 0;
+document.getElementById("submitButton").addEventListener("click", function (event) {
+  const now = Date.now();
+  if (now - lastClickTime < 500) {  // 500ms delay
+    event.preventDefault();
+    alert("Double-click detected! Please click only once.");
+  } else {
+    lastClickTime = now;
+    // Proceed with the action
+  }
+});
+```
+
+✅ Makes it harder for attackers to exploit timing tricks.
+
+✅ Prevents unintended double-clicks on sensitive actions.
+
+#### **4. Trusted Interaction Verification**
+
+**Concept:**  Adds an explicit user confirmation step before executing sensitive actions.
+
+```javascript
+function confirmAction() {
+  let confirmation = confirm("Are you sure you want to proceed?");
+  if (confirmation) {
+    // Perform action
+  }
+}
+```
+
+✅ Requires user intent verification before execution.
+
+✅ Reduces accidental malicious actions.
+
+### **5. Prevent Rapid UI Manipulation Using Mutation Observers**
+
+**Concept:** Monitors for sudden UI changes (such as iframe repositioning) and blocks suspicious activity.
+
+```javascript
+const observer = new MutationObserver(mutations => {
+  mutations.forEach(mutation => {
+    if (mutation.type === "childList" || mutation.type === "attributes") {
+      alert("Suspicious UI changes detected!");
+    }
+  });
+});
+observer.observe(document.body, { childList: true, subtree: true, attributes: true });
+```
+
+✅ Detects hidden UI manipulations in real time.
+
+✅ Prevents attackers from dynamically repositioning elements for Clickjacking attacks.
+
+### **6. Require Intent-Based Gestures for Critical Actions**
+
+**Concept:** Instead of a simple click, users must perform a deliberate gesture, like dragging a slider, before confirming actions.
+
+```html
+<label for="secureAction">Slide to confirm:</label>
+<input type="range" id="secureAction" min="0" max="100" oninput="checkGesture(this.value)">
+<script>
+function checkGesture(value) {
+    if (value == 100) {
+        alert("Confirmed! Action executed.");
+    }
+}
+</script>
+```
```

Comment:
Seriously, a slider on _every_ page? Really? Because  prior to X-FRAME-OPTIONS and CSP frame-ancestors, we had to put the anti-frame busting JavaScript on every bloody page, so I don't see why this would any different. You can't do something that is going to require an explicit action every time a page loads.